### ISSUE TITLE
refactoring IndexedError

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#2809]
`IndexedError` is an interface implemented by errors which belong to a certain slot in a batch. That's important because when a batch fails, the client may need to know which individual command failed.
This currently works because `*proto.Error` stores this information but is prone to bugs since this error isn't obviously linked to the arguments in the batch args any more. For example, consider running a Push which errors; you return this error, but it contains an index which doesn't correspond to the original batch, but to a batch of the PushRequest.
Or #2799, where you prepend and later remove an extra command - failing to adjust the error index means that the error is going to be interpreted as belonging to the wrong slot.

The goal should be to not store the index with an individual error.

* within `Replica`, we can simply return an optional error index
* up from there, everything uses the `Send(BatchRequest) (*BatchResponse, *Error)` interface.
  We change `*Error` to `*ErrorWithIndex`, which is just a non-nullable `Error` and a nullable index (in practice, trivial submessage because ints can't be nullable in proto3). Naming is provisional, we'll probably want to keep the name `Error` for the top-level message.
* `client.{DB,Txn}` needs to expose this appropriately (only use case is `sql.convertBatchError` via `txn.Run(...)`)

**Comment by petermattis** [comment]
@kkaneda, @tschottdorf Anything left to do here?

**Comment by tschottdorf** [comment]
Yes, would still leave it open. The process is ongoing, I'll update later.

**Comment by kkaneda** [comment]
Ah, yes.. This and #3303 requires some more work..

**Comment by kkaneda** [comment]
I'm wondering changing `BatchResponse` to the following makes sense:

```
message ResponseUnionWithError {
  optional ResponseUnion response = 1;
  optional Error error = 2;
}

message BatchResponse {
  optional Header header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
  repeated ResponseUnionWithError responses = 2 [(gogoproto.nullable) = false];
}
```

In addition to the error stored in the header, we will store an error in each response. When i-th request fail, we will store `response` for 0 to (i-1)-th responses, and store `error` for the -th response. I'm wondering this would complicate the code in some places, but we can also avoid maintaining the error index and fixing it in `txn.send`.

**Comment by tschottdorf** [comment]
That was the original code, and we removed that. If a batch fails, you
don't get a response (and even if you got some, they would not actually
have been persisted), so there was a lot more complication (imo) than there
is now.

**Comment by kkaneda** [comment]
Oh, I see. Thanks for the explanation!

**Comment by tamird** [comment]
@tschottdorf @kkaneda can you update this? Not clear on what's left to do.
